hibernate-core/src/test/java/org/hibernate/test/proxy/HibernateUnproxyTest.java

```diff
@@ -22,9 +24,44 @@
 
     @Override
     protected Class<?>[] getAnnotatedClasses() {
-        return new Class<?>[]{Parent.class, Child.class};
+        return new Class<?>[]{Parent.class, Child.class, SubChild.class};
+    }
+
+    @Test
+    public void testInitializedProxySubChildCanBeUnproxied() {
+        Parent u = new Parent();
+        SubChild c = new SubChild();
+        u.setChild(c);
+        doInJPA(this::entityManagerFactory, (entityManager -> {
+            entityManager.persist(u);
+        }));
+        doInJPA(this::entityManagerFactory, (entityManager -> {
+            Parent parent = entityManager.find(Parent.class, u.getId());
+            SubChild unproxiedChild = initializeAndUnproxy(getUser(parent));
+            assertEquals(SubChild.class, unproxiedChild.getClass());
+        }));
+    }
+
+
+    public <U extends Child> U getUser(Parent p) {
+        return (U) p.getChild();
+    }
+
+    public <T> T initializeAndUnproxy(T entity) {
+        if (entity == null) {
+            throw new NullPointerException("Entity passed for initialization is null!");
+        }
+
+        if (!Hibernate.isInitialized(entity)) {
+            Hibernate.initialize(entity);
+            if (entity instanceof HibernateProxy) {
+                entity = (T) ((HibernateProxy) entity).getHibernateLazyInitializer().getImplementation();
+            }
+        }
```

Comment:
Perhaps we should add  `} else { Assert.fail(); }` statements here to make sure that in both scenario's we're dealing with proxies.

Reply:
Done

Reply:
I'll have a try at j9!